include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,384 @@
+/*************************************************************************
+ * Copyright (c) 2017-2019 by the mfmg authors                           *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef MFMG_LANCZOS_LANCZOS_TEMPLATE_HPP
+#define MFMG_LANCZOS_LANCZOS_TEMPLATE_HPP
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <random>
+#include <vector>
+
+#include "lanczos.hpp"
+#include "lanczos_deflatedop.templates.hpp"
+
+// This complex code has to be included before lapacke for the code to compile.
+// Otherwise, it conflicts with boost or Kokkos.
+#include <complex>
+#define lapack_complex_float std::complex<float>
+#define lapack_complex_double std::complex<double>
+#include <lapacke.h>
+
+namespace mfmg
+{
+
+/// \brief Lanczos solver: constructor
+template <typename OperatorType, typename VectorType>
+Lanczos<OperatorType, VectorType>::Lanczos(
+    OperatorType const &op, boost::property_tree::ptree const &params)
+    : _op(op)
+{
+  _is_deflated = params.get<bool>("is_deflated");
+  _maxit = params.get<int>("max_iterations");
+  _tol = params.get<double>("tolerance");
+  _percent_overshoot = params.get<int>("percent_overshoot", 0);
+
+  assert(0 <= _percent_overshoot && _percent_overshoot <= 100);
+
+  if (_is_deflated)
+  {
+    _num_evecs_per_cycle = params.get<int>("num_eigenpairs_per_cycle");
+    _num_cycles = params.get<int>("num_cycles");
+
+    assert(_num_evecs_per_cycle >= 1);
+    assert(_num_cycles >= 1);
+    assert(_maxit >= _num_evecs_per_cycle &&
+           "maxit too small to produce required number of eigenvectors.");
+  }
+  else
+  {
+    _num_requested = params.get<int>("num_eigenpairs");
+
+    assert(_num_requested >= 1);
+    assert(_maxit >= _num_requested &&
+           "maxit too small to produce required number of eigenvectors.");
+  }
+
+  assert(_op.m() == _op.n() && "Operator must be square");
+
+  assert(_maxit >= 0);
```

Comment:
This assertion seems superfluous since you have an assert for `_maxit` in both branches above.